pkg/bytecode/compiler.go

```diff
@@ -259,6 +259,9 @@ func (c *Compiler) compileDecl(decl *parser.Decl) error {
 }
 
 func (c *Compiler) compileAssignment(stmt *parser.AssignmentStmt) error {
+	if err := c.Compile(stmt.Target); err != nil {
```

Comment:
I didn't realise it was a bug to not compile the target of an assignment - I just assumed it was unimplemented because there were no more complex targets than simple variables at the time. So I don't see this as a bug fix but just extending functionality.

Since array and map values are implemented as well as the index operator, I see this change as a separate PR that allows array and map indexing on the left-hand side of an assignment. This slicing PR would follow.

The reason for this split is that 1) this functionality is not really related to slicing, and 2) it really needs more tests to test those forms of assignment. The test added in this commit does not actually test that assigning to an index works - it tests that the assignment does not affect the original value after slicing, but that test can also pass if the index assignment does not work at all. That test can be then added in the slicing PR once index assignment has been merged.

And I don't believe it does work, because the current assignment logic after this commit always creates a new global and that should not happen for array and map index assignment. It also appears that the code in the previous commit does not copy the slice so it would cause mutations in the original slice when modifying a subslice.

Reply:
> I see this change as a separate PR that allows array and map indexing on the left-hand side of an assignment

This would feel incomplete to me without being able to handle dot expressions, when I open the new PR is it reasonable to implement those in it as well?

Reply:
I think that should be fine.

Reply:
@camh- I went back and found that we'd need a new opcode for assigning to an index expression, so I made it part of #318 - I think that dot expressions can follow in another PR